graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/tracing/FederatedTracingInstrumentation.java

```diff
@@ -0,0 +1,282 @@
+
+        if (debuggingEnabled) {
+            tracingMap.put(DEBUG_KEY, trace.toString());
+        }
+
+        return CompletableFuture.completedFuture(new ExecutionResultImpl(executionResult.getData(), executionResult.getErrors(), tracingMap));
+    }
+
+    @Override
+    public InstrumentationContext<Object> beginFieldFetch(InstrumentationFieldFetchParameters parameters) {
+        FederatedTracingState state = parameters.getInstrumentationState();
+
+        long startNanos = System.nanoTime();
+        return whenCompleted((result, throwable) -> {
+            long endNanos = System.nanoTime();
+
+            DataFetchingEnvironment env = parameters.getEnvironment();
+            ExecutionStepInfo executionStepInfo = env.getExecutionStepInfo();
+
+            Node node = state.addNode(executionStepInfo, startNanos, endNanos);
+
+            if (throwable != null) {
+                node.addError(throwable);
+            }
+
+            if (result instanceof DataFetcherResult && ((DataFetcherResult) result).hasErrors()) {
+                DataFetcherResult theResult = (DataFetcherResult) result;
+                List<GraphQLError> errors = theResult.getErrors();
+                errors.forEach(node::addError);
+            }
+        });
+    }
+
+    /**
+     * Stores timing information and a map of nodes for field information.
+     */
+    private class FederatedTracingState implements InstrumentationState {
+        private final Instant startRequestTime;
+        private final long startRequestNanos;
+        private final LinkedHashMap<ExecutionPath, Node> nodeMap;
+
+        private FederatedTracingState() {
+            // record start time when creating instrumentation state for a request
+            startRequestTime = Instant.now();
+            startRequestNanos = System.nanoTime();
+
+            // Create node map with the root node
+            nodeMap = new LinkedHashMap<>();
+            nodeMap.put(ExecutionPath.rootPath(), new Node(ExecutionPath.rootPath()));
+        }
+
+        Reports.Trace toProto() {
+            return Reports.Trace.newBuilder()
+                    .setStartTime(getStartTimestamp())
+                    .setEndTime(getNowTimestamp())
+                    .setDurationNs(getDuration())
+                    .setRoot(buildNodeTree())
+                    .build();
+        }
+
+        /**
+         * Adds node to nodeMap and recursively ensures that all parent node exist.
+         */
+        Node addNode(ExecutionPath path, Node node) {
+            if (!nodeMap.containsKey(path)) {
+                nodeMap.put(path, node);
+            }
+
+            List<Object> pathParts = path.toList();
+            ExecutionPath parentPath = ExecutionPath.fromList(pathParts.subList(0, pathParts.size() - 1));
+
+            if (!nodeMap.containsKey(parentPath)) {
+                addNode(parentPath, new Node(parentPath));
+            }
+
+            nodeMap.get(parentPath).addChild(node);
```

Comment:
oh hm. I didn't realize we were indexing the fields by path **and** creating their children. I suppose that's relatively low memory overhead because we're just dealing with pointers underneath, but I haven't gotten to the point yet that necessitates indexing by field or justifies the value in it.

Reply:
I'm using the same strategy that I found in [this fork](https://github.com/salsify/apollo-tracing-ruby/blob/feature/new-apollo-api/lib/apollo_tracing/tracer.rb) that adds non-federated apollo tracing to the ruby gem. Pretty much just rewrote it in Java here!

You're probably right though, and removing the additional object created for each node would be a performance win. I'll take a look and come back to you if I can't figure out a way to do it!